### DISCUSSION TITLE
Trouble with labels.json Format for Dependency Parser Training in spaCy v3

### DISCUSSION THREAD

**Discussion by LilitKharatyan** [explosion/spaCy#13065] [Help: Coding & Implementations] [answered]
I'm trying to train a dependency parser using spaCy v3 and I've been facing issues with the setup of my labels.json file.
I've gathered all the unique labels from my training data and organized them into a list.
From my understanding, spaCy v3 expects the labels.json to have a structure: Dict[str, Dict[str, Dict[str, int]]].
I first tried to format my labels as:
```
{
    "parser": {
        "labels": {
            "nsubj": 0,
            "root": 1,
            "dobj": 2,
            ...
        }
    }
}
```
However, when referencing this in my config.cfg and starting training, I encounter an error related to the initialization of these labels.
I have also tried to format them in a simple list which works perfectly for the tagger for example, but again I receive an error.

Using the spacy debug config config.cfg command, everything appears to be okay, which suggests that the issue might lie in the way I've formatted my labels.

Questions:

Could anyone clarify the exact expected format of the labels.json for dependency parser training in spaCy v3?
Are there additional considerations I need to account for. 
Any insights would be greatly appreciated!

**Comment by adrianeboyd** [comment, marked as answer]
The labels files aren't intended to be generated by hand, and each pipeline component has its own internal format. You can generate the labels files by using `spacy init labels`. If you have a very large or streamed/infinite corpus, you can generate the labels based on a representative subset of your training data instead of the whole training corpus.

But be aware that in most cases, the labels file isn't required. `spacy train` can determine the labels from the provided training corpus in the initialization step. If you have a large training corpus and you're repeatedly training on the same data, the main benefit is that you save some time at the beginning of `spacy train` by providing the labels.

**Reply by LilitKharatyan** [reply]
Thank you very much for your response. Additionally, I have a hard time finding a recipe, a code (assistance) any relevant information for training a dependency parser for a new language. There is a lot of info for the exciting one but not for training one from scratch. Do you think there is a relevant piece of information? Thank you.

**Reply by adrianeboyd** [reply]
The UD corpora are a good place to get started with training dependency parsers, and there's an example project called `pipelines/tagger_parser_ud`: https://github.com/explosion/projects/tree/v3/pipelines/tagger_parser_ud

To get started, clone it with `spacy project clone pipelines/tagger_parser_ud` and have a look at the projects documentation: https://spacy.io/usage/projects

**Comment by LilitKharatyan** [comment]
Thanks, I have tried this too. While I succeed in training only the tagger, I am unable to train the parser (with the same dataset, btw). After the initialization is done it takes ages for it to start the actual training process (actually it never started). My dataset is around 5000 sentences and tagger finishes training in 7-8 min usually. Where can be the issue?

`=========================== Initializing pipeline ===========================
[2023-10-16 14:43:10,266] [INFO] Set up nlp object from config

[2023-10-16 14:43:10,290] [INFO] Pipeline: ['tok2vec', 'parser']

[2023-10-16 14:43:10,297] [INFO] Created vocabulary

[2023-10-16 14:43:10,301] [INFO] Finished initializing nlp object
`
It never goes beyond this stage.

**Reply by adrianeboyd** [reply]
Which version of spacy exactly?

My first guess is that there is a cycle in one of your dependency trees (so it's not actually a tree) and the parser's projectivization step is stuck in an infinite loop.

If it's a relatively recent version of spacy (v3.4+, I think) there should be a check for cycles that raises an error instead of hanging, so either it's an older version of spacy or there's some edge case that the check isn't catching?

You can try running `spacy debug data`, but it may also hang because it does the same conversion.

One fairly low-tech debugging suggestion: to try to figure out which sentence is the problem, you can override the corpus `limit` setting to gradually increase the number of training sentences: `spacy train ... --corpora.train.limit 500` and then gradually increase it until it hangs. (But be aware that setting `limit` too low lead to other parser problems unless you also decrease `min_action_freq`, related to error `E024`.)

**Reply by adrianeboyd** [reply]
I can also recommend the UD validation tools for checking for various issues like this (I even see a new script that's supposed to break cycles, but I haven't tried it myself): https://github.com/UniversalDependencies/tools

**Reply by LilitKharatyan** [reply]
Thank you very much for the answers. Actually, my dataset is validated with the UD validator, which is why it is surprising that I am getting this cycle. Additionally, I used this code for breaking the cycles (no help again). Let's see how it goes and what their solution might be found!

**Reply by adrianeboyd** [reply]
Sorry this is so frustrating! In the past I've mainly seen this kind of hanging related to cycles, but no one has reported problems for a while, I think especially since we added better checks and errors.

So it sounds more likely that you're running into a different bug, but it's hard for us to debug further unless we can reproduce it on our end. Have you tried training with different subsets of the corpus to try to pinpoint which sentence is causing problems?

**Reply by LilitKharatyan** [reply]
I will try to train with subsets, but is there a way to debug or detect the sets that cause the problem with this information from the error?
**ValueError: Found cycle in dependency graph: [11, 11, 1, 9, 3, 2, 1, 6, 11, 1, 9, 1]**

**Reply by adrianeboyd** [reply]
This is actually pretty annoying to debug because spacy `Doc`s don't have any kind of internal IDs that can be shown in the debugging messages.

I'd have to double-check, but I'm pretty sure that the these heads in the message are 0-indexed while the UD heads are 1-indexed. It's odd that there's no `0` (root) in this list, but it's also possible that the tree has been partially projectivized so the heads no longer correspond to the original UD tree.

You don't have to go through the whole training process, you can just work with the `DocBin` loaded from the `.spacy` file and run something along these lines on each doc:

```
nlp = spacy.blank(lang)
doc_bin = DocBin().from_disk("train.spacy")
for doc in doc_bin.get_docs(nlp.vocab):
    example = Example(nlp.make_doc(doc.text), doc)
    heads, deps = example.get_aligned_parse(projectivize=True)
```

(As a warning, this code is untested but should be pretty close to what you need. You can double-check how this is done in `spacy debug data` to see the same steps. And this doesn't include any tokenizer customization, which may also be a factor.)

**Reply by LilitKharatyan** [reply]
The thing is: I was trying to do it with this basic code:
```
!python -m spacy init config -l xcl --pipeline parser configs/parser_only.cfg
!python -m spacy train config.cfg --output ./output_dir/ --paths.train train.spacy --paths.dev val.spacy
```
But no luck so far. Are there any training recipes, scripts that can be used to train dependency parser (for a new language) and overcome this cycle issue.

P.S. If there are guidelines or information (eval.metrics/ documentation) on releasing new models (we have a morphologizer and POS tagger ready) I will be grateful.

**Reply by adrianeboyd** [reply]
I really do think you'll need to go through your data sentence by sentence to find the problem. If the data is public I can potentially take a quick look in case it is due to a bug in the parser.

But if you have a UD corpus you have quite a few other options. You could also try training, e.g., a Stanza or a UDPipe model.

An easy way to publish/share a new spacy pipeline package is through `spacy-huggingface-hub`, which uploads the packaged `.whl` file and creates a model card with metrics from the pipeline metadata, e.g.: https://huggingface.co/spacy/ca_core_news_lg

**Reply by LilitKharatyan** [reply]
Thank you very much for the answer. This is the [training corpus](https://github.com/UniversalDependencies/UD_Classical_Armenian-CAVaL/blob/master/xcl_caval-ud-test.conllu). I would appreciate any help.

**Reply by adrianeboyd** [reply]
Hmm, this isn't really enough data to train a parser.

Using `hy` as the language the tokenization performance is very bad (f-score 0.69), but the parser does train without hanging.

Steps:

```
spacy convert -T -n 10 xcl_caval-ud-test.conllu .
spacy init config -l hy -p parser config.cfg
spacy train config.cfg --paths.train xcl_caval-ud-test.spacy --paths.dev xcl_caval-ud-test.spacy
```

```
ℹ No output directory provided
ℹ Using CPU

=========================== Initializing pipeline ===========================
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['tok2vec', 'parser']
ℹ Initial learn rate: 0.001
E    #       LOSS TOK2VEC  LOSS PARSER  DEP_UAS  DEP_LAS  SENTS_F  SCORE 
---  ------  ------------  -----------  -------  -------  -------  ------
  0       0          0.00       147.83     8.62     5.50     0.02    0.07
  3     200       2691.53     23636.93    41.33    33.34    52.97    0.37
  6     400       4449.37     15134.16    49.09    44.98    51.49    0.47
  9     600       5288.34     11023.64    52.42    49.69    59.56    0.51
 12     800       5202.59      8311.72    54.16    52.33    67.64    0.53
 15    1000       5380.47      6891.37    54.81    53.49    68.50    0.54
```

(Obviously you don't actually want to use the same train/dev data, but this tests the parser training part with the whole dataset.)

My next guess is that it's something related to the tokenizer, or some interaction with the tokenizer. Is the custom `xcl` code available somewhere?

If the tokenizer performance is this low, nothing else in the pipeline will be able train well because the gold annotation doesn't line up with the predicted tokens. I usually try to get languages well over 0.95 f-score, and 0.95 is still the low end (it's still very noticeable in the pipeline performance), 0.995 is a lot better.

For most UD corpora with spacy, it makes sense to merge subtokens because the spacy tokenizer doesn't include a multi-word token expander. Stanza and other more UD-focused libraries do, though.

**Reply by LilitKharatyan** [reply]
Thank you very much for the response. 
I actually initialize the config file with the xcl code not the 'hy' (modern Armenian) the rest is the same I guess. As for the language code, it is just one file with the basic settings, no example, or stopword file is included. Do you think the problem can be in it?

**Reply by adrianeboyd** [reply]
First see if you can run the same conversion + training as shown above, and then change only the language to see if that makes a difference?

**Reply by LilitKharatyan** [reply]
Thank you very much. I managed to start the training (although the results are not satisfactory even though I used twice as much data as we have released). The only reasonable explanation I find for the error is the way I converted my dataset to the spacy binary format (i used to do it differently and maybe the error arose from it). Thank you for the full support. 
Another question: currently xcl (classical Armenian) is not a language supported by spacy. How can we request an addition of it. What should we provide for that? 
Thank you

**Reply by adrianeboyd** [reply]
To add a new language: https://github.com/explosion/spaCy/discussions/13135#discussioncomment-7616779

**Reply by LilitKharatyan** [reply]
Thanks a lot. I am sorry but I have another question which was never answered in the discussions:
I am training (almost) a full pipeline of models: tokenizer, trainable_lemmatizer, tagger, morphologizer, parser and I see that accuracy differs based on which pipelines are trained together, e.g. I receive the best results when morphologizer and tagger are trained together, while when I add for example parser to this pipeline of training, the results/ accuracy downgrade.
Is it a common behavior, and are there best practices on this issue, or am I doing smth wrong? Thanks
